spacy/kb/candidate.pyx

```diff
@@ -1,74 +1,116 @@
 # cython: infer_types=True, profile=True
 
-from typing import Iterable
+from ..typedefs cimport hash_t
+
 from .kb cimport KnowledgeBase
-from ..tokens import Span
+from .kb_in_memory cimport InMemoryLookupKB
+from ..errors import Errors
 
 cdef class Candidate:
-    """A `Candidate` object refers to a textual mention (`alias`) that may or may not be resolved
-    to a specific `entity` from a Knowledge Base. This will be used as input for the entity linking
+    """A `Candidate` object refers to a textual mention that may or may not be resolved
+    to a specific `entity_id` from a Knowledge Base. This will be used as input for the entity_id linking
     algorithm which will disambiguate the various candidates to the correct one.
-    Each candidate (alias, entity) pair is assigned a certain prior probability.
+    Each candidate (mention, entity_id) pair is assigned a certain prior probability.
 
     DOCS: https://spacy.io/api/kb/#candidate-init
     """
 
-    def __init__(self, KnowledgeBase kb, entity_hash, entity_freq, entity_vector, alias_hash, prior_prob):
-        self.kb = kb
-        self.entity_hash = entity_hash
-        self.entity_freq = entity_freq
-        self.entity_vector = entity_vector
-        self.alias_hash = alias_hash
-        self.prior_prob = prior_prob
+    def __init__(
+        self,
+        entity_id: str,
+        entity_vector: vector[float],
+        prior_prob: float,
+    ):
+        """Initializes properties of abstract base class `Candidate`.
+        entity_id (Union[str, int]): Unique entity ID.
+        entity_vector (List[float]): Entity embedding.
+        prior_prob (float): Prior probability of entity for this mention - i.e. the probability that, independent of
+            the context, this mention resolves to this entity_id in the corpus used to build the knowledge base. In
+            cases in which this isn't always possible (e.g.: the corpus to analyse contains mentions that the KB corpus
+            doesn't) it might be better to eschew this information and always supply the same value.
+        """
+        # Make sure abstract KB is not instantiated.
+        if self.__class__ == Candidate:
+            raise TypeError(
+                Errors.E1046.format(cls_name=self.__class__.__name__)
+            )
 
-    @property
-    def entity(self) -> int:
-        """RETURNS (uint64): hash of the entity's KB ID/name"""
-        return self.entity_hash
+        self._entity_id_ = entity_id
+        # Note that hashing an int value yields the same int value.
+        self._entity_id = hash(entity_id)
```

Comment:
Continuing with the same reasoning. The original thought was to use `vocab` to do the `str`-`int` conversions so we'd only have to store the `int`

Reply:
Why aren't we doing that anymore? `KnowledgeBase` still has a `vocab`?

Reply:
- One motivation behind the design was that we'd support `str` identifier for entities (see above).
- We have `vocab`, but an entity ID isn't necessarily in there IMO. If you have "Q123456789", this might never occur in `vocab`, but be a valid ID. Or am I missing something here?

Reply:
The `vocab` can be extended though, right? That's how it worked before? I guess I'm just confused again why we needed to change this.